xdsl/transforms/loop_invariant_code_motion.py

```diff
@@ -32,9 +41,82 @@ def can_be_hoisted(op: Operation, target_region: Region) -> bool | None:
     for child in op.walk():
         for operand in child.operands:
             operand_owner = operand.owner
-            assert operand_owner is not None
             if op.is_ancestor(operand_owner):
                 continue
             if target_region.is_ancestor(operand_owner):
                 return False
     return True
+
+
+def _move_loop_invariant_code(
+    regions: Sequence[Region],
+    should_move_out_of_region: Callable[[Operation, Region], bool],
+    move_out_of_region: Callable[[Operation, Region], None],
+):
+    num_moved = 0
+    for region in regions:
+        # add top-level operations in the loop body to the worklist
+        worklist = [op for block in region.blocks for op in block.ops]
+
+        while worklist:
+            op = worklist.pop(0)
+            # Skip ops that have already been moved. Check if the op can be hoisted.
+            if op.parent_region() != region:
+                continue
+
+            if not (
+                should_move_out_of_region(op, region) and can_be_hoisted(op, region)
+            ):
+                continue
+
+            move_out_of_region(op, region)
+            num_moved += 1
+
+            # Since the op has been moved, we need to check its users within the
+            # top-level of the loop body.
+
+            for res in op.results:
+                for use in res.uses:
+                    user = use.operation
+                    if user.parent_region() == region:
+                        worklist.append(user)
+
+    return num_moved
+
+
+def _should_move_out_of_region(op: Operation, region: Region) -> bool:
+    return is_side_effect_free(op) and is_speculatable(op)
+
+
+def move_loop_invariant_code(loop: scf.ForOp):
+    loop_regions = (loop.body,)
+
+    builder = Builder(InsertPoint.before(loop))
+
+    def _move_out_of_region(op: Operation, region: Region) -> None:
+        op.detach()
+        builder.insert(op)
+
+    _move_loop_invariant_code(
+        loop_regions, _should_move_out_of_region, _move_out_of_region
+    )
+
+
+class LoopInvariantCodeMotion(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: scf.ForOp, rewriter: PatternRewriter) -> None:
+        move_loop_invariant_code(op)
+
+
+class LoopInvariantCodeMotionPass(ModulePass):
+    """
+    Moves operations without side effects out of loops, provided they do not depend on
+    values defined the loops.
+    """
+
+    name = "licm"
+
+    def apply(self, ctx: Context, op: builtin.ModuleOp) -> None:
+        PatternRewriteWalker(
+            LoopInvariantCodeMotion(), walk_reverse=True, walk_regions_first=True
```

Comment:
Maybe you should explicitely mark it as not recursive? Otherwise it will run multiple times, even though the second time nothing should change AFAIK

Reply:
Will do